File: src/main/java/me/aov/sellgui/addons/AddonManager.java
package me.aov.sellgui.addons;

import me.aov.sellgui.SellGUIMain;
import org.bukkit.Bukkit;
import org.bukkit.plugin.InvalidDescriptionException;
import org.bukkit.plugin.InvalidPluginException;
import org.bukkit.plugin.Plugin;
import org.bukkit.plugin.PluginDescriptionFile;
import org.bukkit.plugin.UnknownDependencyException;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.List;
import java.util.ListIterator;
import java.util.Locale;
import java.util.Optional;
import java.util.jar.JarEntry;
import java.util.jar.JarFile;
import java.util.logging.Level;

public final class AddonManager {
    private final SellGUIMain main;
    private final List<Plugin> loadedAddons = new ArrayList<>();
    private File addonsFolder;

    public AddonManager(SellGUIMain main) {
        this.main = main;
    }

    public void loadAddons() {
        if (!main.getConfig().getBoolean("addons.enabled", true)) {
            main.getLogger().info("Addon loading is disabled.");
            return;
        }

        String folderName = main.getConfig().getString("addons.folder", "addons");
        if (folderName == null || folderName.trim().isEmpty()) {
            folderName = "addons";
        }

        addonsFolder = new File(main.getDataFolder(), folderName.trim());
        if (!addonsFolder.exists() && !addonsFolder.mkdirs()) {
            main.getLogger().warning("Could not create addons folder: " + addonsFolder.getAbsolutePath());
            return;
        }

        File[] addonJars = addonsFolder.listFiles(file ->
                file.isFile() && file.getName().toLowerCase(Locale.ROOT).endsWith(".jar"));
        if (addonJars == null || addonJars.length == 0) {
            if (main.getConfig().getBoolean("general.debug", false)) {
                main.getLogger().info("No addon jars found in " + addonsFolder.getAbsolutePath());
            }
            return;
        }

        Arrays.sort(addonJars, Comparator.comparing(File::getName, String.CASE_INSENSITIVE_ORDER));
        for (File addonJar : addonJars) {
            loadAddon(addonJar);
        }
    }

    public void disableAddons() {
        ListIterator<Plugin> iterator = loadedAddons.listIterator(loadedAddons.size());
        while (iterator.hasPrevious()) {
            Plugin addon = iterator.previous();
            if (addon == null || !addon.isEnabled()) {
                continue;
            }

            try {
                Bukkit.getPluginManager().disablePlugin(addon);
            } catch (Throwable throwable) {
                main.getLogger().log(Level.WARNING, "Failed to disable addon " + addon.getName(), throwable);
            }
        }
        loadedAddons.clear();
    }

    public File getAddonsFolder() {
        return addonsFolder == null ? new File(main.getDataFolder(), "addons") : addonsFolder;
    }

    private void loadAddon(File addonJar) {
        Optional<String> addonName = readPluginName(addonJar);
        if (addonName.isPresent()) {
            Plugin existing = Bukkit.getPluginManager().getPlugin(addonName.get());
            if (existing != null) {
                main.getLogger().info("Addon " + existing.getDescription().getFullName()
                        + " is already loaded, skipping " + addonJar.getName() + ".");
                return;
            }
        }

        try {
            Plugin addon = Bukkit.getPluginManager().loadPlugin(addonJar);
            if (addon == null) {
                main.getLogger().warning("Could not load addon jar: " + addonJar.getName());
                return;
            }

            Bukkit.getPluginManager().enablePlugin(addon);
            loadedAddons.add(addon);
            main.getLogger().info("Loaded addon " + addon.getDescription().getFullName()
                    + " from addons/" + addonJar.getName());
        } catch (UnknownDependencyException exception) {
            main.getLogger().warning("Could not load addon " + addonJar.getName()
                    + " because a dependency is missing: " + exception.getMessage());
        } catch (InvalidPluginException | InvalidDescriptionException exception) {
            main.getLogger().warning("Invalid addon jar " + addonJar.getName() + ": " + exception.getMessage());
        } catch (Throwable throwable) {
            main.getLogger().log(Level.WARNING, "Failed to load addon " + addonJar.getName(), throwable);
        }
    }

    private Optional<String> readPluginName(File addonJar) {
        try (JarFile jarFile = new JarFile(addonJar)) {
            JarEntry pluginYml = jarFile.getJarEntry("plugin.yml");
            if (pluginYml == null) {
                return Optional.empty();
            }

            try (InputStream inputStream = jarFile.getInputStream(pluginYml)) {
                return Optional.of(new PluginDescriptionFile(inputStream).getName());
            }
        } catch (IOException | InvalidDescriptionException ignored) {
            return Optional.empty();
        }
    }
}

File: src/main/kotlin/me/aov/sellgui/addons/AddonManager.kt
package me.aov.sellgui.addons

import me.aov.sellgui.SellGUIMain
import org.bukkit.Bukkit
import org.bukkit.plugin.InvalidDescriptionException
import org.bukkit.plugin.InvalidPluginException
import org.bukkit.plugin.Plugin
import org.bukkit.plugin.PluginDescriptionFile
import org.bukkit.plugin.UnknownDependencyException
import java.io.File
import java.io.IOException
import java.util.Locale
import java.util.jar.JarFile
import java.util.logging.Level

class AddonManager(private val main: SellGUIMain) {
    private val loadedAddons = ArrayList<Plugin>()
    private var addonsFolder: File? = null

    fun loadAddons() {
        if (!main.config.getBoolean("addons.enabled", true)) {
            main.logger.info("Addon loading is disabled.")
            return
        }
        val folderName = main.config.getString("addons.folder", "addons")?.trim().takeUnless { it.isNullOrEmpty() } ?: "addons"
        val folder = File(main.dataFolder, folderName)
        addonsFolder = folder
        if (!folder.exists() && !folder.mkdirs()) {
            main.logger.warning("Could not create addons folder: ${folder.absolutePath}")
            return
        }
        val addonJars = folder.listFiles { file -> file.isFile && file.name.lowercase(Locale.ROOT).endsWith(".jar") }?.sortedBy { it.name.lowercase(Locale.ROOT) } ?: emptyList()
        if (addonJars.isEmpty()) {
            if (main.config.getBoolean("general.debug", false)) main.logger.info("No addon jars found in ${folder.absolutePath}")
            return
        }
        addonJars.forEach(::loadAddon)
    }

    fun disableAddons() {
        loadedAddons.asReversed().forEach { addon ->
            if (addon.isEnabled) try { Bukkit.getPluginManager().disablePlugin(addon) } catch (throwable: Throwable) { main.logger.log(Level.WARNING, "Failed to disable addon ${addon.name}", throwable) }
        }
        loadedAddons.clear()
    }

    fun getAddonsFolder(): File = addonsFolder ?: File(main.dataFolder, "addons")

    private fun loadAddon(addonJar: File) {
        readPluginName(addonJar)?.let { addonName ->
            Bukkit.getPluginManager().getPlugin(addonName)?.let {
                main.logger.info("Addon ${it.description.fullName} is already loaded, skipping ${addonJar.name}.")
                return
            }
        }
        try {
            val addon = Bukkit.getPluginManager().loadPlugin(addonJar)
            if (addon == null) {
                main.logger.warning("Could not load addon jar: ${addonJar.name}")
                return
            }
            Bukkit.getPluginManager().enablePlugin(addon)
            loadedAddons += addon
            main.logger.info("Loaded addon ${addon.description.fullName} from addons/${addonJar.name}")
        } catch (exception: UnknownDependencyException) {
            main.logger.warning("Could not load addon ${addonJar.name} because a dependency is missing: ${exception.message}")
        } catch (exception: InvalidPluginException) {
            main.logger.warning("Invalid addon jar ${addonJar.name}: ${exception.message}")
        } catch (exception: InvalidDescriptionException) {
            main.logger.warning("Invalid addon jar ${addonJar.name}: ${exception.message}")
        } catch (throwable: Throwable) {
            main.logger.log(Level.WARNING, "Failed to load addon ${addonJar.name}", throwable)
        }
    }

    private fun readPluginName(addonJar: File): String? = try {
        JarFile(addonJar).use { jarFile ->
            val pluginYml = jarFile.getJarEntry("plugin.yml") ?: return null
            jarFile.getInputStream(pluginYml).use { PluginDescriptionFile(it).name }
        }
    } catch (_: IOException) {
        null
    } catch (_: InvalidDescriptionException) {
        null
    }
}
